Add unit tests for Linux Datapath Encrypt APIs and Route Rules

jrfastab opened the pull request:
Add additional unit tests for new route rule functionality and ipsec_linux.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6873)
<!-- Reviewable:end -->

jrfastab committed:
cilium: ipsec, route rules unit tests

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: ipsec, add ipsec unit test

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

joestringer commented on pkg/datapath/linux/ipsec/ipsec_linux_test.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+func (p *IPSecSuitePrivileged) TestInvalidLoadKesyFile(c *C) {
+	f, err := os.Create(path)
```
`defer os.RemoveAll(path)`

joestringer replied:
Actually, take a look at my suggestion below about `io.Writer`.

jrfastab replied:
:+1:

joestringer commented on pkg/datapath/linux/ipsec/ipsec_linux_test.go:
```diff
@@ -0,0 +1,150 @@
+
+func (p *IPSecSuitePrivileged) TestLoadKeysNoFile(c *C) {
+	err := LoadIPSecKeysFile(path)
+	c.Assert(err, ErrorMatches, "failed to load IPSec Keys File ipsec_keys_test: open ipsec_keys_test: no such file or directory")
```
nit: Would be nice if this was something like a `c.Assert(os.IsNotExist(err), Equals, true)` just in case the error itself gets reworded at some point.

jrfastab replied:
lgtm.

joestringer replied:
Note: This would require use of `os.PathError{}` in the `LoadIPSecKeysFile()` anywhere we create our own errors, otherwise the type won't be correct and `os.IsNotExist()` needs to understand the type to work properly.

joestringer commented on pkg/datapath/linux/ipsec/ipsec_linux_test.go:
```diff
@@ -0,0 +1,150 @@
+	c.Assert(err, NotNil)
+}
+
+func (p *IPSecSuitePrivileged) TestLoadKesyFile(c *C) {
```
```suggestion
func (p *IPSecSuitePrivileged) TestLoadKeysFile(c *C) {
```

joestringer commented on pkg/datapath/linux/ipsec/ipsec_linux_test.go:
```diff
@@ -0,0 +1,150 @@
+	_, err = f.Write(invalidKeysDat)
+	c.Assert(err, IsNil)
+	f.Sync()
+	err = LoadIPSecKeysFile(path)
```
If `LoadIPSecKeysFile()` takes an `io.Writer` instead of a path (or there's a private variation on the function that does so), then you wouldn't need to write things to the filesystem to test this functionality.

jrfastab replied:
This is a nice idea from the ipsec_linux.go side we can abstract the file out of a base routine so we get, `func loadIPSecKeys(r io.Reader) error`  and then `LoadIPSecKeysFile` opens the file and passes the io.Reader into `loadIPSecKeys`. This makes the unit test independent of the file system we just need to create the "keys" buffer but also it means in the future we can pull keys easily from any io.Reader not just files easily.

joestringer commented on pkg/datapath/linux/ipsec/ipsec_linux_test.go:
```diff
@@ -0,0 +1,150 @@
+	err = UpsertIPSecEndpoint(local, remote)
+	c.Assert(err, IsNil)
+
+	err = DeleteIPSecEndpoint(remote.IP, local.IP)
```
Is there any good test you could put in between the `Upsert()`/`Delete()` that checks for the key or something that will be used from the key, ideally without testing directly the implementation?

jrfastab replied:
We could do a lookup from the netlink side. But I haven't implemented the helper to do that yet. Having a ``FindIPSecEndpoint` helper would be great but I've not actually had any reason to use it except here.

joestringer commented on pkg/datapath/linux/route/route_linux_test.go:
```diff
@@ -119,3 +119,22 @@ func (p *RouteSuitePrivileged) TestReplaceRoute(c *C) {
+func testReplaceRule(c *C, mark, table int) {
+	// delete rule in case it exists from a previous failed run
+	DeleteRule(mark, table)
+	err := ReplaceRule(mark, table)
```
I was thinking "weird, you delete the rule then try to replace... shouldn't it fail?", because we typically name "insert or replace" functions as `UpsertXXX` in other parts of the codebase. Would be nice as a separate followup to make this conform, but not necessary for this PR and not a big deal.

jrfastab replied:
Seems like a good idea there are a few other cleanups like this so I'll add it to the list.

joestringer reviewed (approved):
I have a few notes, the main one below is that if you do a small amount of refactoring then we can unit-test this code without having to deal with the filesystem (or at least test that part separately).

jrfastab committed:
cilium: route, fix deleteRule to include mask and support IPv6

Unit tests found that deleteRule should include a mask to properly
specify delete rule and also that we never remove IPv6 rules. Fix
here.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: ipsec, refactor reading IPSec keys to support io.Reader

Refactor code to read keys from io.Reader with a wrapper to use a file
as the io.Reader. This simplifies the unit tests but also allows us to
read keys from other readers if needed.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

aanm commented on pkg/datapath/linux/ipsec/ipsec_linux_test.go:
```diff
@@ -0,0 +1,140 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
```
needs sign-off

borkmann replied:
Ping @jrfastab, could you add SOB to this one commit so we can merge? Thx

vadorovsky replied:
Ping again ;)

jrfastab replied:
added now. Finally made it to the top of my stack.
